lib/std/os/linux/tls.zig

```diff
@@ -195,124 +192,138 @@ pub fn initTLS() ?*elf.Phdr {
-                l += @sizeOf(DTV);
-                break :blk l;
-            },
-        };
-
-        tls_image = TLSImage{
-            .data_src = @intToPtr([*]u8, phdr.p_vaddr + img_base)[0..phdr.p_filesz],
-            .alloc_size = alloc_size,
-            .tcb_offset = tcb_offset,
-            .dtv_offset = dtv_offset,
-            .data_offset = data_offset,
-            .gdt_entry_number = @bitCast(usize, @as(isize, -1)),
-        };
+    var tls_align_factor: usize = undefined;
+    var tls_data: []const u8 = undefined;
+    if (tls_phdr) |phdr| {
+        tls_align_factor = phdr.p_align;
+        tls_data = @intToPtr([*]u8, img_base + phdr.p_vaddr)[0..phdr.p_memsz];
+    } else {
+        tls_align_factor = @alignOf(*usize);
+        tls_data = &[_]u8{};
     }
 
-    return gnu_stack;
+    // Offsets into the allocated TLS area
+    var tcb_offset: usize = undefined;
+    var dtv_offset: usize = undefined;
+    var data_offset: usize = undefined;
+    var thread_data_offset: usize = undefined;
+    // Compute the total size of the ABI-specific data plus our own control
+    // structures
+    const alloc_size = switch (tls_variant) {
+        .VariantI => blk: {
+            var l: usize = 0;
+            // Unneeded because l is zero
+            // l = mem.alignForward(l, @alignOf(DTV));
+            dtv_offset = l;
+            l += @sizeOf(DTV);
+            l = mem.alignForward(l, @alignOf(CustomData));
+            thread_data_offset = l;
+            l += @sizeOf(CustomData);
+            // Make sure the TP is aligned
+            l = mem.alignForward(l, tls_align_factor);
+            tcb_offset = l;
+            // Ensure there are at least tls_tcb_align_size bytes of padding
+            const min_align = math.max(tls_tcb_align_size, tls_align_factor);
+            l += mem.alignForward(tls_tcb_size, min_align);
+            data_offset = l;
+            l += mem.alignForward(tls_data.len, tls_align_factor);
+            break :blk l;
+        },
+        .VariantII => blk: {
+            var l: usize = 0;
+            data_offset = l;
+            l = mem.alignForward(tls_data.len, tls_align_factor);
+            // The TP is aligned to p_align
+            tcb_offset = l;
+            l += tls_tcb_size;
+            l = mem.alignForward(l, @alignOf(CustomData));
+            thread_data_offset = l;
+            l += @sizeOf(CustomData);
+            l = mem.alignForward(l, @alignOf(DTV));
+            dtv_offset = l;
+            l += @sizeOf(DTV);
+            break :blk l;
+        },
+    };
+
+    tls_image = TLSImage{
+        .data_src = tls_data,
+        .alloc_size = alloc_size,
+        .tcb_offset = tcb_offset,
+        .dtv_offset = dtv_offset,
+        .data_offset = data_offset,
+        .gdt_entry_number = @bitCast(usize, @as(isize, -1)),
+    };
 }
 
-pub fn copyTLS(addr: usize) usize {
-    const tls_img = tls_image.?;
+inline fn alignPtrCast(comptime T: type, ptr: [*]u8) *T {
```

Comment:
I think this function should be manually inlined.

Out of the two call-sites; one doesn't even need the `@alignCast`.

Reply:
> I think this function should be manually inlined.

I made this small helper function as the inline casts produced too much line noise.

> Out of the two call-sites; one doesn't even need the @alignCast.

Both of them go from `*align(1) u8` to `*T` where the latter is ABI-aligned.

Reply:
Line 292 is: `const tcb_ptr = alignPtrCast([*]u8, area.ptr + tls_image.tcb_offset);`
`[*]u8` should have an align of `1`, so `@ptrCast([*]u8, area.ptr + tls_image.tcb_offset)` should be sufficient.

Reply:
> [*]u8 should have an align of 1

It's not. The alignment is either 4 or 8.

Reply:
doh. I was thinking of the child; not the pointer itself.